scania-digital-design-system/sdds
# feat(side-menu): make the collapse event cancelable

**Describe pull-request**  
This PR makes the collapse event cancelable. To do this we emit one event 'sddsCollapse' that can be listened to and canceled by the user. If that event is not canceled we emit an 'internalSddsCollapse' that is listened to by other sdds-side-menu components. 

I also added a onClick event listener to the sdds-side-menu-collapse-button that handles the collapsing automatically if the users chooses to. If they want to control on their own they can cancel the event and toggle the collapased prop. The prop is being watched and emit the same 'internalSddsCollapse' event when it changes.

I also added a propschange event. But since this component only has one prop that is likely to change in runtime (collapsed) I didn't bother with any relevant props for now. I just emit the event, the collapse button listens to it and then checks it's parent again for the new value of collapsed.

In short:

- User can cancel the collapse event by listening for sddsCollapse, being emitted from the collapse-button
- User can still handle collapse by toggling the boolean prop 'collapse' on the sdds-side-menu (would need to prevent the event first though)
- Made the 'Collapse'-text a default slot so the user can translate it.

**NOTE**: please help me battle test the syncing between the sddsCollapse and internalSddsSideMenuPropChange events.

**Solving issue**  
Fixes: https://tegel.atlassian.net/browse/DTS-1105

**How to test**  
1. Go to storybook link below.
2. Check in Components -> Side Menu -> Web Component
3. Check the event being emitted when the collapse-button is clicked.
4. Checkout the branch.
5. Prevent the event and see it fail to collapse.

**Suggested test steps**
